src/QsCompiler/CompilationManager/ContextBuilder.cs

```diff
@@ -534,8 +535,12 @@ internal static IEnumerable<(Range, QsQualifiedName)> CallablesWithContentModifi
             var callables = file.GetCallableDeclarations().Select(tuple => // these are sorted according to their line number
             {
                 var ns = file.TryGetNamespaceAt(tuple.Item2.Start);
-                QsCompilerError.Verify(ns != null, "namespace for callable declaration should not be null"); // invalid namespace names default to an unknown namespace name, but remain included in the compilation
-                return (tuple.Item2.Start, new QsQualifiedName(ns, tuple.Item1));
+
+                // invalid namespace names default to an unknown namespace name, but remain included in the compilation
+                QsCompilerError.Verify(
+                    file.IsNotebook == (ns == null),
+                    "namespace for callable declaration should be null if and only if this is a notebook");
+                return (tuple.Item2.Start, new QsQualifiedName(ns ?? InternalUse.NotebookNamespace, tuple.Item1));
```

Comment:
Is it possible to insert the fake namespace in an earlier stage? It seems like it would be a little nicer if "invalid namespace name" and "implicit notebook namespace" were handled in the same place, instead of notebook namespaces propagating up here.